File: algorithms/medium/rle_iterator.ex
defmodule RLEIterator do
  use GenServer
  
  @spec init_(encoding :: [integer]) :: any
  def init_(encoding) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
    GenServer.call(__MODULE__, {:init, encoding})
  end

  @spec next(n :: integer) :: integer
  def next(n) do
    GenServer.call(__MODULE__, {:next, n})
  end

  def init(encoding) do
    {:ok, encoding}
  end

  def handle_call({:init, encoding}, _from, _state) do
      {:reply, [], encoding}
  end    
  def handle_call({:next, n}, _from, state) do
    {res, new_state} = next_(n, state)
    {:reply, res, new_state}
  end

  defp next_(_n, []), do: {-1, []}
  
  defp next_(n, [ct, num | rest]) do
    if ct >= n do
      {num, [ct - n, num | rest]}
    else
      next_(n - ct, rest)
    end
  end
end

# Your functions will be called as such:
# RLEIterator.init_(encoding)
# param_1 = RLEIterator.next(n)

# RLEIterator.init_ will be called before every test case, in which you can do some necessary initializations.

File: algorithms/medium/rle_iterator.rs
struct RLEIterator {
    encoding: Vec<i32>,
    curr: usize,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RLEIterator {

    fn new(encoding: Vec<i32>) -> Self {
        Self {
            encoding,
            curr: 0,
        }
    }
    
    fn next(&mut self, n: i32) -> i32 {
        if self.curr >= self.encoding.len() {
            return -1;
        }
        self.encoding[self.curr] -= n;
        if self.encoding[self.curr] < 0 {
            let m = -self.encoding[self.curr];
            self.curr += 2;
            return self.next(m);
        }
        self.encoding[self.curr + 1]
    }
}

/**
 * Your RLEIterator object will be instantiated and called as such:
 * let obj = RLEIterator::new(encoding);
 * let ret_1: i32 = obj.next(n);
 */
